### DISCUSSION TITLE
如何让某个Field字段聚焦，尝试使用onFocus触发无效果

### DISCUSSION THREAD

**Discussion by li9269391** [alibaba/formily#1615] [Q&A 2.x] [answered]
测试demo: https://codesandbox.io/s/boring-waterfall-vx5yb?file=/App.tsx
想要实现让某个字段聚焦，通常用于初始化聚焦或配合键盘事件处理业务场景
![image](https://user-images.githubusercontent.com/10287877/122843247-974a0180-d331-11eb-8180-87a41fc1ebad.png)

**Comment by liuweiGL** [comment]
`field` 是一个 model（管理数据/状态的），onFocus 只是一个 hook，用于更新 field 的属性与触发动作（比如校验）。
而你的需求是 UI 层面的，这个需要你使用 ref 来操作 dom。

**Reply by li9269391** [reply]
使用了schema，问题是怎么获取里面的UI层ref？

**Reply by li9269391** [reply]
请问 formily shcema 方式下，有什么思路或方式可以处理这个场景吗？还是说不支持呢？

**Comment by liuweiGL** [comment, marked as answer]
给 Input 组件一个 autoFocus 属性，然后使用:

```ts
{
    name: 'test',
    type: 'string',
    'x-component': 'Input',
    'x-component-props': {
        autoFocus: true
  }
}
```

**Reply by li9269391** [reply]
@liuweiGL 感谢，原来是DOM原属性，但我尝试在某个点击按钮中让它聚焦，给它加 ref 后控制，似乎页面就会卡住了
```
"x-component-props": {
    // autoFocus: true
    ref: inputRef
}
```

**Comment by liuweiGL** [comment]
用 react 的开发思维来思考，点击按钮的时候更新 `x-component-props` 中的 `autoFocus`  属性即可。

**Reply by li9269391** [reply]
尝试了，好像不可以
```
const field = normalForm.query("username").take();
field.setComponentProps({
  // readOnly: true,  // 有效
  // disabled: true,  // 有效
  // 无效
  autoFocus: true,
});
```

**Comment by liuweiGL** [comment]
你使用的什么 UI 库？自己看文档啊，醉了。。。

**Reply by li9269391** [reply]
> 你使用的什么 UI 库？自己看文档啊，醉了。。。

Antd呢

**Reply by li9269391** [reply]
antd 通常是 usrRef 来处理，这边是 schema 用了 field.setComponentProps，就是设置属性的。
有可能是状态改变，组件刷新，又失焦了

**Reply by li9269391** [reply]
执行field.setComponentProps({autoFocus: true,});后看了一下 DOM节点，autoFocus属性是加上去了，但不聚焦，我自己再看看吧。

**Reply by janryWang** [reply]
自己纯dom操作吧，document.querySelector找到input，触发它的focus就行了

**Reply by dufengjiao** [reply]
> 执行field.setComponentProps({autoFocus: true,});后看了一下 DOM节点，autoFocus属性是加上去了，但不聚焦，我自己再看看吧。

请问解决了么，我也遇到这个问题

**Reply by Nokecy** [reply]
autoFocus 只是初始有用的.

**Reply by li9269391** [reply]
> > 执行field.setComponentProps({autoFocus: true,});后看了一下 DOM节点，autoFocus属性是加上去了，但不聚焦，我自己再看看吧。
> 
> 请问解决了么，我也遇到这个问题

模型不处理 DOM，需要自己做一层业务封装操作 DOM，document.querySelector 找到 input
